File: Linked List/reorderLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head) -> None:
        # [0, n-1, 1, n-2, 2, n-3, ...]
        
        # find the middle of the linked list
        # to partition them
        slow = fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # beginning of the second half of the LL
        second = slow.next
        prev = slow.next = None # split the LL

        # reverse the second half of the LL
        while second:
            tmp = second.next
            second.next = prev
            prev = second
            second = tmp

        # merge the two halves
        # prev = second half of the LL
        first, second = head, prev
        while second:
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first, second = tmp1, tmp2

File: Linked List/reorderLinkedList.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reorderList(head *ListNode) {
	var slow, fast *ListNode = head, head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	second := slow.Next
	var prev *ListNode = nil
	slow.Next = nil

	for second != nil {
		var temp *ListNode = second.Next
		second.Next = prev
		prev = second
		second = temp
	}

	var first *ListNode = head
	second = prev

	for second != nil {
		var temp1 *ListNode = first.Next
		var temp2 *ListNode = second.Next

		first.Next = second
		second.Next = temp1

		first = temp1
		second = temp2
	}
}
